Index: mf2b-1.1/Makefile
===================================================================
--- mf2b-1.1.orig/Makefile	2014-01-14 13:50:09.000000000 +1000
+++ mf2b-1.1/Makefile	2014-08-07 17:49:02.315597393 +1000
@@ -3,10 +3,10 @@
 # maximum number of subpatterns recognised
 MAX_SUBS := 16
 
-CC := gcc
-CFLAGS := -Wall -Werror -pedantic -std=c99 -Os \
+CC ?= gcc
+CFLAGS += -Wall -Werror -pedantic -std=c99 -Os \
           -DMAX_SUBS=${MAX_SUBS} -DVERSION="\"${VERSION}\""
-INSTALL := install
+INSTALL ?= install $(if $(STRIP),--strip-program="$(STRIP)")
 
 ifneq (${DEBUG},)
 CFLAGS += -DDEBUG -g
@@ -21,8 +21,9 @@
 	-rm -f ${MF2B_OBJ} mf2b.o mf2b
 
 install: mf2b
-	${INSTALL} -D -s -m 0755 mf2b ${DESTDIR}/usr/sbin/mf2b
-	${INSTALL} -D -m 0644 mf2b.conf ${DESTDIR}/etc/mf2b.conf
+	${INSTALL} -D -m 0755 mf2b ${DESTDIR}/usr/sbin/mf2b
+#	${INSTALL} -D -m 0644 mf2b.conf ${DESTDIR}/etc/mf2b.conf
+install-man:
 	${INSTALL} -D -m 0644 mf2b.8 \
 		${DESTDIR}/usr/share/man/man8/mf2b.8
 	${INSTALL} -D -m 0644 mf2b.conf.5 \
